includes/classes/Indexable/Comment/QueryIntegration.php

```diff
@@ -0,0 +1,336 @@
+		if ( Utils\is_indexing() ) {
+			return;
+		}
+
+		// Add header
+		add_action( 'pre_get_comments', array( $this, 'action_pre_get_comments' ), 5 );
+
+		// Filter comment query
+		add_filter( 'comments_pre_query', [ $this, 'maybe_filter_query' ], 10, 2 );
+	}
+
+	/**
+	 * Add EP header
+	 *
+	 * @param  WP_Comment_Query $query Query object
+	 * @since  3.1
+	 * @return void
+	 */
+	public function action_pre_get_comments( WP_Comment_Query $query ) {
+		if ( ! Indexables::factory()->get( 'comment' )->elasticpress_enabled( $query ) || apply_filters( 'ep_skip_comment_query_integration', false, $query ) ) {
+			return;
+		}
+
+		if ( ! headers_sent() ) {
+			/**
+			 * Manually setting a header as $wp_query isn't yet initialized
+			 * when we call: add_filter('wp_headers', 'filter_wp_headers');
+			 */
+			header( 'X-ElasticPress-Search: true' );
+		}
+	}
+
+	/**
+	 * If WP_Comment_Query meets certain conditions, query results from ES
+	 *
+	 * @param  array            $results Query results.
+	 * @param  WP_Comment_Query $query   Current query.
+	 * @since  3.1
+	 * @return array
+	 */
+	public function maybe_filter_query( $results, WP_Comment_Query $query ) {
+		$this->indexable = Indexables::factory()->get( 'comment' );
+
+		if ( ! $this->indexable->elasticpress_enabled( $query ) || apply_filters( 'ep_skip_comment_query_integration', false, $query ) ) {
+			return $results;
+		}
+
+		$new_comments = apply_filters( 'ep_wp_query_cached_comments', null, $query );
+
+		if ( null !== $new_comments ) {
+			return $new_comments;
+		}
+
+		$formatted_args = $this->indexable->format_args( $query->query_vars );
+
+		$scope = 'current';
+		if ( ! empty( $query->query_vars['sites'] ) ) {
+			$scope = $query->query_vars['sites'];
+		}
+
+		/**
+		 * Filter search scope
+		 *
+		 * @since 3.1
+		 *
+		 * @param mixed $scope The search scope. Accepts `all` (string), a single
+		 *                     site id (int or string), or an array of site ids (array).
+		 */
+		$scope = apply_filters( 'ep_comment_search_scope', $scope );
+
+		if ( ! defined( 'EP_IS_NETWORK' ) || ! EP_IS_NETWORK ) {
+			$scope = 'current';
+		}
+
+		$this->index = null;
+
+		if ( 'all' === $scope ) {
+			$this->index = $this->indexable->get_network_alias();
+		} elseif ( is_numeric( $scope ) ) {
+			$this->index = $this->indexable->get_index_name( (int) $scope );
+		} elseif ( is_array( $scope ) ) {
+			$this->index = [];
+
+			foreach ( $scope as $site_id ) {
+				$this->index[] = $this->indexable->get_index_name( $site_id );
+			}
+
+			$this->index = implode( ',', $this->index );
+		}
+
+		$ep_query = $this->indexable->query_es( $formatted_args, $query->query_vars, $this->index );
+
+		if ( false === $ep_query ) {
+			$query->elasticsearch_success = false;
+			return $results;
+		}
+
+		if ( ! empty( $query->query_vars['count'] ) ) {
+			return count( $ep_query['documents'] );
+		}
+
+		$query->found_comments        = $ep_query['found_documents'];
```

Comment:
`$query->max_num_pages` should be calculated here as well.
```suggestion
		$query->found_comments        = $ep_query['found_documents'];
		$query->max_num_pages         = $query->query_vars['number'] <= 0 ? 1 : max( 1, ceil( $query->found_comments / absint( $query->query_vars['number'] ) ) );
```